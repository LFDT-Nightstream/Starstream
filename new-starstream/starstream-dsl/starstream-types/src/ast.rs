//! Abstract syntax tree for the pared-down Starstream DSL.

use chumsky::span::SimpleSpan;
use serde::Serialize;

/// Entire program: a sequence of statements.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Program {
    pub statements: Vec<Statement>,
}

/// Statements allowed by the current grammar.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Statement {
    VariableDeclaration {
        name: Identifier,
        value: Spanned<Expr>,
    },
    Assignment {
        target: Identifier,
        value: Spanned<Expr>,
    },
    If {
        condition: Spanned<Expr>,
        then_branch: Block,
        else_branch: Option<Block>,
    },
    While {
        condition: Spanned<Expr>,
        body: Block,
    },
    Block(Block),
    Expression(Spanned<Expr>),
}

/// `{ statement* }`
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Block {
    pub statements: Vec<Statement>,
}

/// Expressions with precedence encoded via parser, not the AST shape.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Expr {
    Literal(Literal),
    Identifier(Identifier),
    Unary {
        op: UnaryOp,
        expr: Box<Spanned<Expr>>,
    },
    Binary {
        op: BinaryOp,
        left: Box<Spanned<Expr>>,
        right: Box<Spanned<Expr>>,
    },
    Grouping(Box<Spanned<Expr>>),
}

#[derive(Clone, Debug, Serialize, PartialEq)]
pub enum Literal {
    Integer(i64),
    Boolean(bool),
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub enum UnaryOp {
    Negate,
    Not,
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq)]
pub enum BinaryOp {
    Multiply,
    Divide,
    Remainder,
    Add,
    Subtract,
    Less,
    LessEqual,
    Greater,
    GreaterEqual,
    Equal,
    NotEqual,
    And,
    Or,
}

#[derive(Clone, Debug, Serialize, PartialEq, Eq, Hash)]
pub struct Identifier {
    pub name: String,
    #[serde(skip)]
    pub span: Option<SimpleSpan>,
}

impl Identifier {
    pub fn new(name: impl Into<String>, span: Option<SimpleSpan>) -> Self {
        Self {
            name: name.into(),
            span,
        }
    }
}

/// Helper for attaching source spans to AST nodes.
#[derive(Clone, Debug, Serialize, PartialEq)]
pub struct Spanned<T> {
    pub node: T,
    #[serde(skip)]
    pub span: SimpleSpan,
}

impl<T> Spanned<T> {
    pub fn new(node: T, span: SimpleSpan) -> Self {
        Self { node, span }
    }

    pub fn none(node: T) -> Spanned<T> {
        Spanned {
            node,
            span: SimpleSpan {
                start: 0,
                end: 0,
                context: (),
            },
        }
    }
}
