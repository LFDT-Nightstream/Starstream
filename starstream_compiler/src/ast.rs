//! AST types describing a Starstream source file.

#[derive(Clone, Debug, Default)]
pub struct StarstreamProgram {
    pub items: Vec<ProgramItem>,
}

#[derive(Clone, Debug)]
pub enum ProgramItem {
    Utxo(Utxo),
    Script(Script),
    Token(Token),
}

#[derive(Clone, Debug)]
pub struct Utxo {
    pub name: Identifier,
    pub items: Vec<UtxoItem>,
}

#[derive(Clone, Debug)]
pub enum UtxoItem {
    Abi(Abi),
    Main(Main),
    Impl(Impl),
    Storage(Storage),
}

#[derive(Clone, Debug)]
pub struct Main {
    pub type_sig: Option<OptionallyTypedBindings>,
    pub block: Block,
}

#[derive(Clone, Debug)]
pub struct Token {
    pub name: Identifier,
    pub items: Vec<TokenItem>,
}

#[derive(Clone, Debug)]
pub enum TokenItem {
    Bind(Bind),
    Unbind(Unbind),
    Abi(Abi),
    Mint(Mint),
}

#[derive(Clone, Debug)]
pub struct Bind(pub Block);

#[derive(Clone, Debug)]
pub struct Unbind(pub Block);

#[derive(Clone, Debug)]
pub struct Mint(pub Block);

#[derive(Clone, Debug)]
pub struct Impl {
    pub name: Identifier,
    pub definitions: Vec<FnDef>,
}

#[derive(Clone, Debug)]
pub struct Script {
    pub definitions: Vec<FnDef>,
}

#[derive(Clone, Debug)]
pub struct Storage {
    pub bindings: TypedBindings,
}

#[derive(Clone, Debug)]
pub struct Sig {
    pub name: Identifier,
    pub input_types: Vec<Type>,
    pub output_type: Option<Type>,
}

#[derive(Clone, Debug)]
pub struct FnSig(pub Sig);

#[derive(Clone, Debug)]
pub struct FnDef {
    pub name: Identifier,
    pub inputs: OptionallyTypedBindings,
    pub output: Option<Type>,
    pub body: Block,
}

#[derive(Clone, Debug)]
pub enum EffectSig {
    EffectSig(Sig),
    EventSig(Sig),
    ErrorSig(Sig),
}

#[derive(Clone, Debug)]
pub enum AbiElem {
    FnSig(FnSig),
    EffectSig(EffectSig),
}

#[derive(Clone, Debug)]
pub struct Abi {
    pub values: Vec<AbiElem>,
}

#[derive(Clone, Debug)]
pub struct Identifier(pub String);

#[derive(Clone, Debug)]
pub enum Type {
    BaseType(Identifier, Option<Vec<Type>>),
    Object(TypedBindings),
    FnType(TypedBindings, Option<Box<Type>>),
}

#[derive(Clone, Debug)]
pub enum Statement {
    BindVar {
        var: Identifier,
        value: Expr,
        mutable: bool,
    },
    Return(Option<Expr>),
    Resume(Option<Expr>),
    Assign(Identifier, Expr),
    With(Block, Vec<(Effect, Block)>),
    While(Expr, LoopBody),
    Loop(LoopBody),
}

#[derive(Clone, Debug)]
pub enum LoopBody {
    Statement(Box<Statement>),
    Block(Block),
    Expr(Expr),
}

#[derive(Clone, Debug)]
pub enum Expr {
    Equals(Box<Self>, Box<Self>),
    NotEquals(Box<Self>, Box<Self>),
    LessThan(Box<Self>, Box<Self>),
    Add(Box<Self>, Box<Self>),
    Or(Box<Self>, Box<Self>),
    Sub(Box<Self>, Box<Self>),
    Pow(Box<Self>, Box<Self>),
    Mul(Box<Self>, Box<Self>),
    And(Box<Self>, Box<Self>),
    Div(Box<Self>, Box<Self>),
    Neg(Box<Self>),
    PrimaryExpr(
        PrimaryExpr,
        Option<Arguments>,
        Vec<(Identifier, Option<Arguments>)>,
    ),
    BlockExpr(BlockExpr),
}

#[derive(Clone, Debug)]
pub enum BlockExpr {
    IfThenElse(Box<Expr>, Box<Block>, Option<Box<Block>>),
    Block(Block),
}

#[derive(Clone, Debug)]
pub enum PrimaryExpr {
    Null,
    Number(f64),
    Bool(bool),
    Ident(Identifier),
    ParExpr(Box<Expr>),
    Yield(Box<Expr>),
    Raise(Box<Expr>),
    Object(Type, Vec<(Identifier, Expr)>),
}

#[derive(Clone, Debug)]
pub enum ExprOrStatement {
    Expr(Expr),
    Statement(Statement),
}

#[derive(Clone, Debug)]
pub enum Block {
    Chain {
        head: Box<ExprOrStatement>,
        tail: Box<Block>,
    },
    Close {
        semicolon: bool,
    },
}

#[derive(Clone, Debug)]
pub struct Arguments {
    pub xs: Vec<Expr>,
}

#[derive(Clone, Debug)]
pub struct OptionallyTypedBindings {
    pub values: Vec<(Identifier, Option<Type>)>,
}

#[derive(Clone, Debug)]
pub struct TypedBindings {
    pub values: Vec<(Identifier, Type)>,
}

#[derive(Clone, Debug)]
pub struct Effect {
    pub ident: Identifier,
    pub type_sig: OptionallyTypedBindings,
}
