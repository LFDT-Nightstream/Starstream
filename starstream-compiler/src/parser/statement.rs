use chumsky::{prelude::*, recursive::recursive};
use starstream_types::ast::{Block, Statement};

use super::{context::Extra, expression, primitives};

/// Parser for individual statements.
pub fn parser<'a>() -> impl Parser<'a, &'a str, Statement, Extra<'a>> {
    recursive(|statement_parser| {
        let block_parser = statement_parser
            .repeated()
            .collect::<Vec<_>>()
            .delimited_by(just('{').padded(), just('}').padded())
            .map(|statements| Block { statements })
            .boxed();

        parser_with_block(block_parser).boxed()
    })
}

pub(super) fn parser_with_block<'a>(
    block_parser: impl Parser<'a, &'a str, Block, Extra<'a>> + Clone + 'a,
) -> impl Parser<'a, &'a str, Statement, Extra<'a>> {
    let expr = expression::parser().boxed();

    let variable_declaration = just("let")
        .padded()
        .ignore_then(primitives::identifier())
        .then_ignore(just('=').padded())
        .then(expr.clone())
        .then_ignore(just(';').padded())
        .map(|(name, value)| Statement::VariableDeclaration { name, value });

    let assignment = primitives::identifier()
        .then_ignore(just('=').padded())
        .then(expr.clone())
        .then_ignore(just(';').padded())
        .map(|(target, value)| Statement::Assignment { target, value });

    let if_statement = just("if")
        .padded()
        .ignore_then(
            expr.clone()
                .delimited_by(just('(').padded(), just(')').padded()),
        )
        .then(block_parser.clone())
        .then(
            just("else")
                .padded()
                .ignore_then(block_parser.clone())
                .or_not(),
        )
        .map(|((condition, then_branch), else_branch)| Statement::If {
            condition,
            then_branch,
            else_branch,
        });

    let while_statement = just("while")
        .padded()
        .ignore_then(
            expr.clone()
                .delimited_by(just('(').padded(), just(')').padded()),
        )
        .then(block_parser.clone())
        .map(|(condition, body)| Statement::While { condition, body });

    let block_statement = block_parser.clone().map(Statement::Block).padded();

    let expression_statement = expr
        .then_ignore(just(';').padded())
        .map(Statement::Expression);

    choice((
        variable_declaration,
        assignment,
        if_statement,
        while_statement,
        block_statement,
        expression_statement,
    ))
    .boxed()
}

#[cfg(test)]
mod tests {
    use super::*;
    use indoc::indoc;

    macro_rules! assert_statement_snapshot {
        ($code:expr) => {{
            let stmt = parser()
                .parse(indoc! { $code })
                .into_result()
                .expect("statement should parse");

            insta::with_settings!({
                description => format!("Code:\n\n{}", indoc! { $code }),
                omit_expression => true,
                prepend_module_to_snapshot => true,
            }, {
                insta::assert_debug_snapshot!(stmt);
            });
        }};
    }

    #[test]
    fn let_binding() {
        assert_statement_snapshot!("let answer = 42;");
    }

    #[test]
    fn assignment() {
        assert_statement_snapshot!("value = value + 1;");
    }

    #[test]
    fn if_else() {
        assert_statement_snapshot!(
            r#"
            if (flag) { let a = 1; } else { value = value + 1; }
            "#
        );
    }

    #[test]
    fn while_loop() {
        assert_statement_snapshot!(
            r#"
            while (count < 10) { count = count + 1; }
            "#
        );
    }

    #[test]
    fn block() {
        assert_statement_snapshot!(
            r#"
            { let a = 1; a = a + 1; }
            "#
        );
    }

    #[test]
    fn expression_statement() {
        assert_statement_snapshot!("value + 1;");
    }
}
