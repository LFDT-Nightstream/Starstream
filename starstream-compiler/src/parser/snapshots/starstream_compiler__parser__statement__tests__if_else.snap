---
source: starstream-compiler/src/parser/statement.rs
description: "Code:\n\nif (flag) { let a = 1; } else { value = value + 1; }\n"
---
If {
    condition: Spanned {
        node: Identifier(
            Identifier {
                name: "flag",
                span: Some(
                    4..8,
                ),
            },
        ),
        span: 4..8,
    },
    then_branch: Block {
        statements: [
            VariableDeclaration {
                name: Identifier {
                    name: "a",
                    span: Some(
                        16..17,
                    ),
                },
                value: Spanned {
                    node: Literal(
                        Integer(
                            1,
                        ),
                    ),
                    span: 20..21,
                },
            },
        ],
    },
    else_branch: Some(
        Block {
            statements: [
                Assignment {
                    target: Identifier {
                        name: "value",
                        span: Some(
                            32..37,
                        ),
                    },
                    value: Spanned {
                        node: Binary {
                            op: Add,
                            left: Spanned {
                                node: Identifier(
                                    Identifier {
                                        name: "value",
                                        span: Some(
                                            40..45,
                                        ),
                                    },
                                ),
                                span: 40..46,
                            },
                            right: Spanned {
                                node: Literal(
                                    Integer(
                                        1,
                                    ),
                                ),
                                span: 48..49,
                            },
                        },
                        span: 40..49,
                    },
                },
            ],
        },
    ),
}
